src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp

```diff
@@ -1717,12 +1716,36 @@ void ShenandoahHeap::set_concurrent_strong_root_in_progress(bool in_progress) {
   }
 }
 
-void ShenandoahHeap::set_concurrent_weak_root_in_progress(bool in_progress) {
-  if (in_progress) {
-    _concurrent_weak_root_in_progress.set();
-  } else {
-    _concurrent_weak_root_in_progress.unset();
+void ShenandoahHeap::set_concurrent_weak_root_in_progress(bool cond) {
+  set_gc_state_mask(WEAK_ROOTS, cond);
+}
+
+class ShenandoahDisableWeakRootsClosure : public HandshakeClosure {
+public:
+  inline ShenandoahDisableWeakRootsClosure() : HandshakeClosure("ShenandoahDisableWeakRoots") {}
+  inline void do_thread(Thread* thread) {
+    char gc_state = ShenandoahThreadLocalData::gc_state(thread);
+    gc_state &= ~((char)ShenandoahHeap::WEAK_ROOTS);
+    ShenandoahThreadLocalData::set_gc_state(thread, gc_state);
   }
+};
+
+void ShenandoahHeap::disable_concurrent_weak_root_in_progress_concurrently() {
+
+  // We need a rendezvous here to avoid the following race:
+  // 1. Java thread reads referent, sees non-null but unreachable oop
+  // 2. GC thread clears the referent
```

Comment:
How is this possible? GC threads (workers) are not running when rendezvous roots. What do I miss?

Reply:
But they are running *before* rendezvous roots.

Reply:
Okay, looks good.